### ISSUE TITLE
Invalid state transition. Details: {"transition":"MERGED","state":"READY"} 

### ISSUE THREAD

**Issue by savinskiy-konstantin** [deepstreamIO/deepstream.io#1078]
I've faced with the following error for some records during trying to update record:
```
Invalid state transition.
Details: {"transition":"MERGED","state":"READY"} 
History: 
	From - to LOADING_OFFLINE via -
	From LOADING_OFFLINE to SUBSCRIBING via 23
	From SUBSCRIBING to READY via 4

StateMachine.transition             @ state-machine.ts:37
RecordCore.onRecordRecovered        @ record-core.ts:767
eval                                @ merge-strategy-service.ts:75
exports.REMOTE_WINS                 @ merge-strategy.ts:8
MergeStrategyService.merge          @ merge-strategy-service.ts:74
RecordCore.recoverRecordFromMessage @ record-core.ts:726
RecordCore.RecordHandler            @ record-core.ts:480
RecordHandler.handle                @ record-handler.ts:411
eval                                @ connection.ts:323
Connection.onMessages               @ connection.ts:297
socket.onmessage                    @ socket-factory.ts:42
```
Code is simple:
```
const { DeepstreamClient } = window.DeepstreamClient;
const client = new DeepstreamClient(url, { path: '' } );

client.login((success) => {
  if (success) {
    client.record.getRecord('user').whenReady((record) => {
      record.set('test', 1);
    });
  }
});
```

There are many following messages in pm2 error logs:
```
VERSION_EXISTS | open tried to update record user to version 19633 but it already was 19632
```

Server version: `5.1.1`
Client version: `5.1.4`

What is the reason for the error?
How can the error be fixed?

Thank you

**Comment by jaime-ez** [comment]
Hi, I can not reproduce the issue based on the code you share. Seems to me you might have the offline functionality enabled and thus the record version from the client and server differ?

**Comment by yasserf** [comment]
The error itself is due to probably missing a state transition in the record handler, since it is finite and we need to output the states ourselves. That being said, going from ready to merged and skipping merging state seems odd.

My only guess is either you have other code running we are not aware about, since a single write can’t produce more than one conflict, or something managed to enter an infinite loop to produce many conflict logs. Given the simplicity of the code I feel like I would support the former.

Eitherways this issue can’t be resolved on our side without better reporting / logs or best of all, reproducible code.

**Comment by jaime-ez** [comment]
Closing this for now. Please re-open if issue persists.

**Comment by zoltangbereczky** [comment]
I have the very same issue using mongodb as storage.

**Comment by savinskiy-konstantin** [comment]
I got back to this problem. 

> Seems to me you might have the offline functionality enabled and thus the record version from the client and server differ?

offlineEnabled is disabled by default.

We have a simple architecture. The client creates/updates their own record and subscribed to others. But sometimes with some records happens described above issue.
We can not detect steps to reproduce. 

Can you help with troubleshooting? I see these "broken" records. What I can check to detect what is wrong with this record and why I cannot update the record?

Thanks

**Comment by yasserf** [comment]
So the error is being thrown due to a lack of a transition here:

https://github.com/deepstreamIO/deepstream.io-client-js/blob/master/src/record/record-core.ts#L933

So my guess is you might be updating the records while the client is offline or if another process/client updates it and there's weird state. 

The actual issue seems to be here:

https://github.com/deepstreamIO/deepstream.io-client-js/blob/master/src/record/record-core.ts#L530

Where we are recovering a message but we haven't put the record into a recovering state. We need to transition using this:

https://github.com/deepstreamIO/deepstream.io-client-js/blob/master/src/record/record-core.ts#L956

So basically:

```
    if (message.action === RECORD_ACTION.VERSION_EXISTS) {
      this.stateMachine.transition(RECORD_ACTION.VERSION_EXISTS)
      this.recoverRecordFromMessage(message as RecordWriteMessage)
      return
    }
```

Not sure about the side effects, but that should fix this specific issue.

**Comment by savinskiy-konstantin** [comment]
I've added your code but it does not help. I've gets two errors instead of one.
![image](https://user-images.githubusercontent.com/10757583/103350684-77840000-4aa9-11eb-9d18-31fe2ed11c4f.png)

**Comment by yasserf** [comment]
Could be because we are using the wrong actions, can you try this line?

      this.stateMachine.transition(RECORD_OFFLINE_ACTIONS.VERSION_EXISTS)

**Comment by savinskiy-konstantin** [comment]
The error in the console is gone with this line but the record doesn't update anyway.

**Comment by yasserf** [comment]
awesome. That issue is because your merge strategy is remote wins.

exports.REMOTE_WINS                 @ merge-strategy.ts:8

You need to set merge strategies to LOCAL_WINS if you expect them to work.

Generally worth looking into your code to see if your updating records when the connection is offline/not yet ready. But eitherways that should be a solution.

**Comment by savinskiy-konstantin** [comment]
Thanks.

I will review cases when the connection is offline or not yet ready.

What about LOCAL_WINS, it doesn't help.

I cannot set merge strategy for connection how described in docs https://deepstream.io/tutorials/core/datasync/handling-data-conflicts/. Please clarify this for the new client version.

I've tried the following code:

```
const { DeepstreamClient } = window.DeepstreamClient;
client = new DeepstreamClient(serverUrl, { path: '' } );

client.login({}, function () {
    var currentUserRecord = client.record.getRecord('name');
    currentUserRecord.whenReady(function (record) {
        record.setMergeStrategy(function(localValue, localVersion, remoteValue, remoteVersion, callback) {
            callback(null, localValue);
        })
        record.set('test', 1);
    });
});
```

**Comment by savinskiy-konstantin** [comment]
I've been able to reproduce the issue by following steps.
1. Code with updating record by interval
```
var recordName = 'test';
const { DeepstreamClient } = window.DeepstreamClient;
client = new DeepstreamClient(serverUrl, { path: '' } );

client.login({}, function () {
    var currentUserRecord = client.record.getRecord(recordName);
    currentUserRecord.whenReady(function (record) {
        var updateRecord = function() {
            var currentdate = new Date();
            var datetime = currentdate.getMinutes() + ":" + currentdate.getSeconds();
            record.set('time', currentdate, err => {
                if (err) {
                    console.log('Record set with error:', err)
                } else {
                    console.log('Record set without error')
                }
            });
        }
        var updateRecordInterval = setInterval(updateRecord, 1000);
    });
});
```
2. Disconnect the connection (disable internet, etc.)
3. Then I get the error in the console and cannot update the record anymore

I think this case should be handled by Deepstream Client

**Comment by yasserf** [comment]
thank you for the code! yeah that should be covered.

I'm currently a bit busy on another deadline, but will pick this up as soon as I have time.

**Comment by rom-melnik** [comment]
+1

**Comment by yasserf** [comment]
I will not have time to do this as I'm working more in the capacity of PR reviewer for now. If anyone feels like fixing this in deepstream I'll be happy to review it.

**Comment by jaime-ez** [comment]
Hi @savinskiy-konstantin I still can't reproduce the issue. This is my setup:

server.js using server v5.2.2
```
const { Deepstream } = require('@deepstream/server')

const server = new Deepstream({})
server.start()
```

index.html
```
<!DOCTYPE html>
<html>
  <head>
    <script src="https://cdn.jsdelivr.net/npm/@deepstream/client@5.2.1/dist/bundle/ds.min.js"></script>
  </head>
  <body>
    <input type="text" />
    <script type="text/javascript">
      window.onload = function () {
        //js goes here
        const { DeepstreamClient } = window.DeepstreamClient
        const client = new DeepstreamClient('ws://192.168.0.12:6020/deepstream')
        const recordName = 'test'
        client.login({ username: '1' }, () => {
          var currentUserRecord = client.record.getRecord(recordName);
          currentUserRecord.whenReady(function (record) {
              var updateRecord = function() {
                  var currentdate = new Date();
                  var datetime = currentdate.getMinutes() + ":" + currentdate.getSeconds();
                  record.set('time', currentdate, err => {
                      if (err) {
                          console.log('Record set with error:', err)
                      } else {
                          console.log('Record set without error')
                      }
                  });
              }
          var updateRecordInterval = setInterval(updateRecord, 1000);
          })
        })
      }
    </script>
  </body>
</html>
```

Then I go to index.html on the browser, and see the record without error on the console. Disable wifi, no logs to the console, and if enabled in less than a minute, it reconnects and logs success. else it will log client offline, and after reconnecting continues to log success.

Please let me know which steps you took in order to reproduce the error, cheers

**Comment by max-rudnev** [comment]
Hello, everyone.

Faced with the same (a similar) issue.
Managed to reproduce it using the code from the @savinskiy-konstantin's [comment](https://github.com/deepstreamIO/deepstream.io/issues/1078#issuecomment-752923415), but:
1. with a smaler interval of 10 ms;
2. and running the code in two browser tabs.

Once the code ran in the second tab the following error occurs:
![Screenshot_1](https://user-images.githubusercontent.com/5476222/110756911-7282c000-8253-11eb-99bd-4176e689959e.png)

**Comment by jaime-ez** [comment]
I think the issue is here: https://github.com/deepstreamIO/deepstream.io-client-js/blob/master/src/record/record-core.ts#L845
the stateMachine is already in ready state, we already recovered the record, changin it to:
```
    if (utils.deepEquals(oldValue, remoteData)) {
      if (this.stateMachine.state !== RECORD_STATE.READY) {
        this.stateMachine.transition(RECORD_OFFLINE_ACTIONS.MERGED)
      }
      return
    }
```
has worked for me.

Can you check if it works for you ?

**Comment by max-rudnev** [comment]
Hello, everyone.

@jaime-ez, thank you for your response.
Unfortunately, your solution doesn't help in my [case](https://github.com/deepstreamIO/deepstream.io/issues/1078#issuecomment-796559451).
It is still reproducible with the modification.

Also, in my case, the error occurs after transiting the state on this line: https://github.com/deepstreamIO/deepstream.io-client-js/blob/master/src/record/record-core.ts#L851.

**Comment by jaime-ez** [comment]
I was testing based on the case you mentioned. I think it's weird you're seing the error on the line you mention, since the `onRecordRecovered` function should be called once, sync the record, and then subsequent calls should exit on the line I mentioned before since the record would be already recovered...can you add the if block I proposed where you see the error ocurring ?

**Comment by max-rudnev** [comment]
> can you add the if block I proposed where you see the error ocurring and see what happens?

Retested the case.
The error occurs without either of the try blocks.
It doesn't occur with both the try blocks only:
![Screenshot_1](https://user-images.githubusercontent.com/5476222/111278453-362fd500-8642-11eb-835f-723059b4e994.png)

**Comment by jaime-ez** [comment]
Well then put the if block once on line 844 (before both cases tested) and that should cover all out of sync cases. Let us know how it goes

**Comment by max-rudnev** [comment]
I've retested the case.
Now the following issue occurs:
![Screenshot_1](https://user-images.githubusercontent.com/5476222/111602318-e5ec7a80-87db-11eb-8513-73588b8700ed.png)
![Screenshot_2](https://user-images.githubusercontent.com/5476222/111602345-eb49c500-87db-11eb-93be-4ae267b0f54e.png)
when transiting the state on this line:
![Screenshot_3](https://user-images.githubusercontent.com/5476222/111602695-48de1180-87dc-11eb-970a-4ed68426fc2c.png)

**Comment by yasserf** [comment]
i'll take a look at this this weekend, seems quite a few people are running into it

**Comment by yasserf** [comment]
So using the code samples above I found the issue if we were transitioning from ready to merged which is invalid. So since the data is already the same (meaning nothing actually changed) I just put a guard to say if it isn't in the merging state then just ignore it.

Hopefully that fixes it, in 5.2.2

**Comment by max-rudnev** [comment]
@yasserf, thank you for your work.
Please notice that in my case the error also occurs in case of non-equal data (please see the [comment](https://github.com/deepstreamIO/deepstream.io/issues/1078#issuecomment-800063363)).
Therefore, this case is still reproducible with 5.2.2.

**Comment by yasserf** [comment]
Can you please add a code segement that replicates the issue (eg: https://github.com/deepstreamIO/deepstream.io-client-js/blob/master/bugs/1078-mergeconflict.ts) and using that we can put in a fix. Thanks

**Comment by max-rudnev** [comment]
I reproduce the issue on your code with a smaller interval of 10 ms.

The full code:
```<!DOCTYPE html>
<html>
    <head>
        <script src="/js/ds.js"></script>
    </head>
    <body>
        <script type="text/javascript">
          const test = () => {
            var recordName = 'test';

            const { DeepstreamClient } = window.DeepstreamClient;
            client = new DeepstreamClient('wss://<hostname>:6020/', { path: '' });

            client.login({}, function () {
              const currentUserRecord = client.record.getRecord(recordName)
              currentUserRecord.whenReady(function (record) {
                var updateRecord = function () {
                  currentUserRecord.set({data: Math.random()}, err => {
                    if (err) {
                      console.log('Record set with error:', err)
                    } else {
                      console.log('Record set without error')
                    }
                  });
                }
                setInterval(updateRecord, 10);
              });
            });
          };

          if (window.addEventListener) {
            window.addEventListener('load', test, false);
          } else if (window.attachEvent) {
            window.attachEvent('onload', test);
          }
        </script>
    </body>
</html>

**Comment by jaime-ez** [comment]
Ok then, I think I got it:

> Well then put the if block once on line 844 (before both cases tested) and that should cover all out of sync cases. Let us know how it goes

This is correct for eliminating the incorrect state transition that originated this issue: `{"transition":"MERGED","state":"READY"}`

Now for the second error described by @max-rudnev `{"transition":"MERGED","state":"RESUBSCRIBING"}` the thing is that the resubscription to the record coming back from the disconnection is handled via the bulkSubscriptionService [here](https://github.com/deepstreamIO/deepstream.io-client-js/blob/master/src/record/record-core.ts#L416)

Since the default `subscriptionInterval` is 100 ms, thus if we are updating with an interval less than 100 ms will have a "race condition" that generates that invalid state transition. This can be fixed by initialising the client with the option `subscriptionInterval: 0`.

Please test and let us knwo if it works.[

**Comment by yasserf** [comment]
hey!

I put in a fix in 5.2.3 that basically stops invalid merge conflicts from happening. However as correctly stated by @jaime-ez, the issue is due to a race condition we don't have much control with. If your updating things at a rate of 10milliseconds a second with an async merge conflict handler we just drop merge conflicts that happen concurrently. Coming up with a better solution is overkill + it will record on the next update anyways.

Hopefully that solves this issue. It will log a warning so that you can be aware how often it happens. My advice would be not to send updates quicker than a single screen refresh frame (or ideally a few) if you want to avoid that behavior.

**Comment by jaime-ez** [comment]
Closing for now, please reopen if other cases arise.
